src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/AuditUtility.cs

```diff
@@ -0,0 +1,287 @@
+
+            if (allVulnerabilityData.Exceptions != null)
+            {
+                ReplayErrors(allVulnerabilityData.Exceptions);
+            }
+
+            if (allVulnerabilityData.KnownVulnerabilities != null)
+            {
+                CheckPackageVulnerabilities(allVulnerabilityData.KnownVulnerabilities);
+            }
+        }
+
+        private void ReplayErrors(AggregateException exceptions)
+        {
+            foreach (Exception exception in exceptions.InnerExceptions)
+            {
+                string messageText = string.Format(Strings.Error_VulnerabilityDataFetch, exception.Message);
+                RestoreLogMessage logMessage = RestoreLogMessage.CreateError(NuGetLogCode.NU1900, messageText);
+                _logger.Log(logMessage);
+            }
+        }
+
+        private void CheckPackageVulnerabilities(IReadOnlyList<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>> knownVulnerabilities)
+        {
+            Dictionary<PackageIdentity, Dictionary<PackageVulnerabilityInfo, List<string>>>? packagesWithKnownVulnerabilities =
+                FindPackagesWithKnownVulnerabilities(knownVulnerabilities);
+
+            if (packagesWithKnownVulnerabilities != null)
+            {
+                // no-op checks DGSpec hash, which means the order of everything must be deterministic.
+                // .NET Framework and .NET Standard don't have Deconstructor methods for KeyValuePair
+                foreach (var kvp1 in packagesWithKnownVulnerabilities.OrderBy(p => p.Key.Id))
+                {
+                    PackageIdentity package = kvp1.Key;
+                    Dictionary<PackageVulnerabilityInfo, List<string>> vulnerabilities = kvp1.Value;
+                    foreach (var kvp2 in vulnerabilities.OrderBy(v => v.Key.Url.OriginalString))
+                    {
+                        PackageVulnerabilityInfo vulnerability = kvp2.Key;
+                        List<string> affectedGraphs = kvp2.Value;
+                        (string severityLabel, NuGetLogCode logCode) = GetSeverityLabelAndCode(vulnerability.Severity);
+                        string message = string.Format(Strings.Warning_PackageWithKnownVulnerability,
+                            package.Id,
+                            package.Version.ToNormalizedString(),
+                            severityLabel,
+                            vulnerability.Url);
+                        RestoreLogMessage restoreLogMessage =
+                            RestoreLogMessage.CreateWarning(logCode,
+                            message,
+                            package.Id,
+                            affectedGraphs.OrderBy(s => s).ToArray());
+                        restoreLogMessage.ProjectPath = _projectFullPath;
+                        _logger.Log(restoreLogMessage);
+                    }
+                }
+            }
+        }
+
+        private static List<PackageVulnerabilityInfo>? GetKnownVulnerabilities(
```

Comment:
Is it worth having the methods in this class return read-only versions of the collections?
```suggestion
        private static ReadOnlyCollection<PackageVulnerabilityInfo>? GetKnownVulnerabilities(
```

Reply:
`List` overrides `GetEnumerator` to return a struct enumerator, so `foreach` loops do not allocate. If we switch to `ReadOnlyCollection`, then it will return `IEnumerable`, which means it will allocate.  Since restore is our main perf hot path, I  was trying to minimize allocations. If the method was public, then I'd probably agree, but it's private, so I think it's reasonable to optimize for perf.